check/src/unify_type.rs

```diff
@@ -1177,7 +1189,7 @@ impl<'a, 'e> Unifier<State<'a>, ArcType> for UnifierState<'a, Subsume<'e>> {
             }
 
             (_, &Type::Forall(_, _, Some(_))) => {
-                let r = r.instantiate_generics(&mut FnvMap::default());
```

Comment:
This was the actual unsound line. Unfortunately fixing this (and the direct fallout) exposed a unfication ambiguity where I have two sound programs but only one or the other will successfully typecheck depending on some minor tweaks to the typechecker... Researching options for now.

Reply:
See the unify_with_inferred_forall_in_record test https://github.com/gluon-lang/gluon/pull/606/files#diff-697e0a05f79902051453dccb6fd941c2R1050